### PR TITLE
Eve - Completed Tokenizing Input and Up Time

### PR DESCRIPTION
**Pull request by EveCoulibaly**
I still have one error for the UpTime code as I have no integrated the Version Check mini project.

### CODE REVIEW

File: src/codeu/chat/util/Tokenizer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    public Tokenizer (String source) {
+        this.source = source;
+        at = 0;
+    }
+
+    public String next() throws IOException {
+        // Skip all leading whitespace
+        while (remaining() > 0 && Character.isWhitespace(peek())) {
+            read();  // ignore the result because we already know that it is a whitespace character
```
Comment (serenadipity) [inline src/codeu/chat/util/Tokenizer.java:21]: In-line comments are nice and clear. They are not too frequent and they make the code very readable.

File: src/codeu/chat/util/Tokenizer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+
+    private char read() throws IOException {
+        final char c = peek();
+        at += 1;
+        return c;
+    }
+
+    private String readWithNoQuotes() throws IOException {
```
Comment (serenadipity) [inline src/codeu/chat/util/Tokenizer.java:55]: Good function names (easy to understand)

File: test/codeu/chat/util/TokenizerTest.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        assertEquals(tokenizer.next(), "hello");
+        assertEquals(tokenizer.next(), "world");
+        assertEquals(tokenizer.next(), "how");
+        assertEquals(tokenizer.next(), "are");
+        assertEquals(tokenizer.next(), "you");
+        assertEquals(tokenizer.next(), null);
+    }
+
+    @Test
+    public void testWithNoQuotes() throws IOException {
```
Comment (serenadipity) [inline test/codeu/chat/util/TokenizerTest.java:20]: The code is well designed here. Tests are broken up into separate functions with descriptive names. You could include a few more unit tests in this file just to test more complex edge cases.

File: src/codeu/chat/client/core/View.java [hunk 4]
```diff
@@ -126,14 +127,38 @@ public View(ConnectionSource source) {
...
         messages.addAll(Serializers.collection(Message.SERIALIZER).read(connection.in()));
-      } else {
+      }
+      else {
         LOG.error("Response from server failed.");
       }
-    } catch (Exception ex) {
+    }
+    catch (Exception ex) {
       System.out.println("ERROR: Exception during call on server. Check log for details.");
```
Comment (serenadipity) [inline src/codeu/chat/client/core/View.java:136]: Error messages are nice and clear.

File: src/codeu/chat/common/NetworkCode.java [hunk 5]
```diff
@@ -35,5 +35,7 @@
       RELAY_READ_REQUEST = 27,
       RELAY_READ_RESPONSE = 28,
       RELAY_WRITE_REQUEST = 29,
-      RELAY_WRITE_RESPONSE = 30;
+      RELAY_WRITE_RESPONSE = 30,
```
Comment (serenadipity) [inline src/codeu/chat/common/NetworkCode.java:38]: Variable names are appropriately descriptive

File: src/codeu/chat/client/core/Context.java [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
 public final class Context {
 
-  private final BasicView view;
+  private final View view;
```
Comment (serenadipity) [inline src/codeu/chat/client/core/Context.java:29]: Code is well-designed for the system (the code is broken up into classes appropriately)

File: src/codeu/chat/client/commandline/Panel.java [hunk 7]
```diff
@@ -48,10 +49,13 @@ public void register(String commandName, Command command) {
   // will be returned. True will be return if a command is found. Whether or not
   // the command was successful is not returned.
   //
-  public boolean handleCommand(String commandName, Scanner line) {
```
Comment (serenadipity) [inline src/codeu/chat/client/commandline/Panel.java:51]: Since this change modifies how the command from the command line is parsed and passed to the command interface, maybe a brief comment should be inserted to clarify the change made to this function.

File: src/codeu/chat/client/commandline/Chat.java [hunk 8]
```diff
@@ -47,9 +54,22 @@ public Chat(Context context) {
   //
   public boolean handleCommand(String line) {
 
-    final Scanner tokens = new Scanner(line.trim());
+    final List<String> args = new ArrayList<String>();
+    final Tokenizer tokenizer = new Tokenizer(line);
+
+    try {
```
Comment (serenadipity) [inline src/codeu/chat/client/commandline/Chat.java:60]: This is an example of an exception that you can include in your test cases.

File: src/codeu/chat/util/Tokenizer.java (hunk 1 above)
Comment (EveCoulibaly) [reply src/codeu/chat/util/Tokenizer.java:21]: Yes! I need to be better about commenting.

File: .idea/CodeU-Summer-2017.iml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (nissimnatanov) [inline .idea/CodeU-Summer-2017.iml:1]: Please see my comment to Jonathan re skipping .idea files using .gitignore.

File: src/codeu/chat/client/commandline/Chat.java [hunk 10]
```diff
@@ -148,6 +168,19 @@ public void invoke(Scanner args) {
...
 
+    panel.register("info", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        final ServerInfo info = context.getInfo();
+        if (info == null) {
+          System.out.println("ERROR: Invalid information");
+        } else {
+          System.out.println("Server info: version: " + NetworkCode.SERVER_INFO_RESPONSE + " start time: " + info.startTime.toString());
+          // Print the server info to the user in a pretty way
```
Comment (nissimnatanov) [inline src/codeu/chat/client/commandline/Chat.java]: comments usually come before the subject code :)

File: src/codeu/chat/common/ServerInfo.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+
+package codeu.chat.common;
+
+import codeu.chat.util.Time;
+
+public final class ServerInfo {
+
+    public final Time startTime;
+
+    public ServerInfo() {
```
Comment (nissimnatanov) [inline src/codeu/chat/common/ServerInfo.java]: when you have two constructors, please comment in which scenario each of the constructors is used.

File: src/codeu/chat/util/Tokenizer.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        read(); // read the closing the quote that allowed us to exit the loop
+        return token.toString();
+    }
+
+
+
+    private StringBuilder token;
+    private String source;
+    private int at;
```
Comment (nissimnatanov) [inline src/codeu/chat/util/Tokenizer.java]: two comments here:
* it is generally good to keep members that are supposed to stay const as final. please mark source as such
* keep final and non-final fields grouped
* please move fields to the top of the class.

**Review by nissimnatanov — CHANGES_REQUESTED** [review]
